pkg/lib/oauth/handler/handler_authz.go

```diff
@@ -606,3 +640,60 @@ func (h *AuthorizationHandler) generateCodeResponse(
 	resp.Code(code)
 	return nil
 }
+
+func (h *AuthorizationHandler) generateSettingsActionResponse(
+	redirectURI string,
+	idpSessionID string,
+	authenticationInfo authenticationinfo.T,
+	idTokenHintSID string,
+	r protocol.AuthorizationRequest,
+	authz *oauth.Authorization,
+	resp protocol.AuthorizationResponse,
+) error {
+	code, _, err := h.SettingsActionGrantService.CreateSettingsActionGrant(&CreateSettingsActionGrantOptions{
+		Authorization:        authz,
+		IDPSessionID:         idpSessionID,
+		AuthenticationInfo:   authenticationInfo,
+		IDTokenHintSID:       idTokenHintSID,
+		RedirectURI:          redirectURI,
+		AuthorizationRequest: r,
+	})
+	if err != nil {
+		return err
+	}
+
+	resp.Code(code)
+	return nil
+}
+
+func (h *AuthorizationHandler) prepareErrInvalidOAuthResponse(req *http.Request, oauthError protocol.OAuthProtocolError) httputil.Result {
+	resultErr := authorizationResultError{
+		Response:    oauthError.Response,
+		RedirectURI: nil,
+	}
+
+	// Only redirect if oauth session is expired / not found
+	// It mostly happens when user refresh the page or go back to the page after authenication
+	if oauthError.Type() != "invalid_request" {
+		return resultErr
+	}
+
+	redirectURI, err := url.Parse(req.URL.Query().Get("redirect_uri"))
+	if err != nil {
+		return resultErr
+	}
+
+	client := h.ClientResolver.ResolveClient(req.URL.Query().Get("client_id"))
+	if client == nil {
+		return resultErr
+	}
+
+	err = validateRedirectURI(client, h.HTTPProto, h.HTTPOrigin, h.AppDomains, redirectURI)
+	if err != nil {
+		return resultErr
+	}
+
+	resultErr.RedirectURI = redirectURI
```

Comment:
In the places where prepareErrInvalidOAuthResponse is called, redirectURI is already parsed. So prepareErrInvalidOAuthResponse should take the parsed redirect URI instead of *http.Request.